blueocean-personalization/src/main/js/components/DashboardCards.jsx

```diff
@@ -0,0 +1,223 @@
+            return -endCompare;
+        }
+    }
+
+    const startTimeA = favoriteA && favoriteA.item && favoriteA.item.latestRun && favoriteA.item.latestRun.startTime;
+    const startTimeB = favoriteB && favoriteB.item && favoriteB.item.latestRun && favoriteB.item.latestRun.startTime;
+
+    if (startTimeA && startTimeB) {
+        const startCompare = startTimeA.localeCompare(startTimeB);
+
+        if (startCompare !== 0) {
+            return -startCompare;
+        }
+    }
+
+    const queuedTimeA = favoriteA && favoriteA.item && favoriteA.item.latestRun && favoriteA.item.latestRun.enQueueTime;
+    const queuedTimeB = favoriteB && favoriteB.item && favoriteB.item.latestRun && favoriteB.item.latestRun.enQueueTime;
+
+    if (queuedTimeA && queuedTimeB) {
+        const queueCompare = queuedTimeA.localeCompare(queuedTimeB);
+
+        if (queueCompare !== 0) {
+            return -queueCompare;
+        }
+    }
+
+    return 0;
+};
+
+/**
+ * Extract elements from a path string deliminted with forward slashes
+ * @param path
+ * @param begin
+ * @param end
+ * @returns {string}
+ */
+const extractPath = (path, begin, end) => {
+    try {
+        return path.split('/').slice(begin, end).join('/');
+    } catch (error) {
+        return '';
+    }
+};
+
+/**
+ */
+export class DashboardCards extends Component {
+
+    constructor(props) {
+        super(props);
+
+        this.fetchUserInProgress = false;
+        this.fetchFavoritesInProgress = false;
+    }
+
+    componentWillMount() {
+        this._initialize(this.props);
+    }
+
+    componentWillReceiveProps(props) {
+        this._initialize(props);
+    }
+
+    _initialize(props) {
+        const config = this.context.config;
+        const { user, favorites } = props;
+
+        if (user) {
+            this.fetchUserInProgress = false;
+        }
+
+        if (favorites) {
+            this.fetchFavoritesInProgress = false;
+        }
+
+        if (config) {
+            const shouldFetchUser = !user && !this.fetchUserInProgress;
+            const shouldFetchFavorites = user && !favorites && !this.fetchFavoritesInProgress;
+
+            if (shouldFetchUser) {
+                this.fetchUserInProgress = true;
+                this.props.fetchUser(config);
+            }
+
+            if (shouldFetchFavorites) {
+                this.fetchFavoritesInProgress = true;
+                this.props.fetchFavorites(config, user);
+            }
+        }
+    }
+
+    render() {
+        if (!this.props.favorites) {
+            return null;
+        }
+
+        const sortedFavorites = this.props.favorites.sort(sortComparator);
+
+        const favoriteCards = sortedFavorites.map(fav => {
+            const pipeline = fav.item;
+            const latestRun = pipeline.latestRun;
+
+            let fullName;
+            let pipelineName;
+            let branchName;
+
+            if (pipeline._class === 'io.jenkins.blueocean.service.embedded.rest.BranchImpl') {
```

Comment:
really need to centralize this somehow, i think. and maybe use the classes/capabilities api